packages/SwingSet/test/upgrade/test-upgrade.js

```diff
@@ -634,3 +656,52 @@ test('failed upgrade - unknown options', async t => {
   // t.regex(e.message, /upgrade\(\) received unknown options: bad/);
   t.regex(e.message, /upgrade\(\) received unknown options: \(a string\)/);
 });
+
+test('failed vatAdmin upgrade - bad replacement code', async t => {
+  const config = {
+    includeDevDependencies: true, // for vat-data
+    bootstrap: 'bootstrap',
+    defaultReapInterval: 'never',
+    vats: {
+      bootstrap: { sourceSpec: bfile('bootstrap-upgrade.js') },
+    },
+    bundles: {
+      ulrik1: { sourceSpec: bfile('vat-ulrik-1.js') },
+    },
+  };
+
+  const hostStorage = provideHostStorage();
+  await initializeSwingset(config, [], hostStorage);
+  const c = await makeSwingsetController(hostStorage);
+  c.pinVatRoot('bootstrap');
+  await c.run();
+
+  const run = async (method, args = []) => {
+    assert(Array.isArray(args));
+    const kpid = c.queueToVatRoot('bootstrap', method, args);
+    await c.run();
+    const status = c.kpStatus(kpid);
+    const capdata = c.kpResolution(kpid);
+    return [status, capdata];
+  };
+
+  const badVABundle = await bundleSource(
+    new URL('./vat-junk.js', import.meta.url).pathname,
+  );
+  const bundleID = await c.validateAndInstallBundle(badVABundle);
+  const kpid = c.upgradeStaticVat('vatAdmin', false, bundleID, {});
```

Comment:
Let's also use this test to confirm that a failed upgrade (with the same `pause` enabled as we expect to use this in real life) doesn't somehow brick the vat and prevent us from making a successful attempt later. I'm thinking:

* change that `false` to `true`, to trigger pausing
* after the `buildV1` succeeds below:
  * do a `restartVatAdminVat(c)`
  * do another does-it-still-work operation (maybe `buildV2`)

Reply:
Oh, actually putting `false` in the call to `upgradeStaticVat` there was just a straight up mistake on my part.  It should be `true`.  The reason it worked is because there was no other activity going on in the vatAdmin vat at the time (on account of this test being minimalistic as tests generally should be) so we benefitted from a de facto paused state even though we hadn't asked for it.  As a rule it's an error to upgrade a vatAdmin vat without pausing it first; it's just that there's no place to stand to detect this condition, since user code never sees the `stopVat` request -- it just gets shot in the head.  In a more ideal world the `shouldPauseFirst` flag should not be a discretionary parameter for whoever calls `upgradeStaticVat`; rather, the vatAdmin's `upgradeStaticVat` method should be able to find out if the target vat supports `pauseService` and send the message if it does.  Hmm.  I wonder if we should _always_ send `pauseService` and just catch and ignore the "no such method" error if one is thrown.  Can we even do that?  That seems a little sketchy to me, because it makes the method name magic.

I think the `buildV1` after the failed upgrade is sufficient, because we already have lots of tests that restarting vatAdmin between various operations leaves it in a correct state.